pkg/datapath/linux/probes/probes.go

```diff
@@ -217,3 +223,60 @@ func (p *ProbeManager) GetHelpers(prog string) map[string]struct{} {
 	}
 	return nil
 }
+
+// writeHeaders executes bpftool to generate BPF feature C macros and then
+// writes them to the given writer.
+func (p *ProbeManager) writeHeaders(featuresFile io.Writer) error {
+	cmd := exec.WithTimeout(
+		defaults.ExecTimeout, "bpftool", "feature", "probe", "macros")
+	stdoutPipe, err := cmd.StdoutPipe()
+	if err != nil {
+		return fmt.Errorf(
+			"could not initialize stdout pipe for bpftool: %s", err)
+	}
+	stderrPipe, err := cmd.StderrPipe()
+	if err != nil {
+		return fmt.Errorf(
+			"could not initialize stderr pipe for bpftool: %s", err)
+	}
+	if err := cmd.Start(); err != nil {
+		return fmt.Errorf("could not start bpftool: %s", err)
+	}
+
+	writer := bufio.NewWriter(featuresFile)
+	defer writer.Flush()
+
+	go io.Copy(writer, stdoutPipe)
+	if err := cmd.Wait(); err != nil {
+		stderr, err := ioutil.ReadAll(stderrPipe)
+		if err != nil {
+			return fmt.Errorf(
+				"bpftool did not run succesfully and stdout could not be read: %s",
+				err)
+		}
+		return fmt.Errorf("bpftool did not run successfully: %s: %s",
+			err, stderr)
+	}
+	return nil
+}
+
+// CreateHeadersFile creates a C header file with macros indicating which BPF
+// features are available in the kernel.
+func (p *ProbeManager) CreateHeadersFile() error {
+	globalsDir := option.Config.GetGlobalsDir()
+	if err := os.MkdirAll(globalsDir, defaults.StateDirRights); err != nil && os.IsExist(err) {
+		return fmt.Errorf("Could not create runtime directory %s", globalsDir)
+	}
+	featuresFilePath := filepath.Join(globalsDir, "bpf_features.h")
+	featuresFile, err := os.Create(featuresFilePath)
+	if err != nil {
+		return fmt.Errorf("could not create features header file %s: %s",
+			featuresFilePath, err)
+	}
+	defer featuresFile.Close()
+
+	if err := p.writeHeaders(featuresFile); err != nil {
```

Comment:
I might have missed it, but it doesn't look like you preserve the file's header and footer, i.e., `#ifndef BPF_FEATURES_H_` & co.

Reply:
Thanks for pointing that out! Fixed, I added header and footer in `writeHeaders`.